File: server/models/Coupon.js
const mongoose = require("mongoose");

const couponSchema = new mongoose.Schema(
  {
    code: { type: String, required: true, unique: true },
    description: { type: String, required: true },
    discountType: { type: String, enum: ["percentage", "fixed"], required: true },
    discountValue: { type: Number, required: true },
    minPurchase: { type: Number, default: 0 },
    maxDiscount: { type: Number },
    startDate: { type: Date, required: true },
    endDate: { type: Date, required: true },
    isActive: { type: Boolean, default: true },
    usageLimit: { type: Number, default: 1 },
    usedCount: { type: Number, default: 0 },
    restaurantId: { type: String, ref: "Restaurant" },
    // 특정 메뉴 항목에만 적용되는 쿠폰인 경우
    applicableItems: [{ type: String }],
  },
  { timestamps: true }
);

module.exports = mongoose.model("Coupon", couponSchema);

File: flutter_client/lib/app/data/models/coupon.dart
// file: lib/app/data/models/coupon.dart
import 'package:uuid/uuid.dart';
import 'package:get/get.dart';

class Coupon {
  final String id;
  final String code;
  final String description;
  final String discountType;
  final double discountValue;
  final double minPurchase;
  final double? maxDiscount;
  final DateTime startDate;
  final DateTime endDate;
  final bool isActive;
  final int usageLimit;
  final int usedCount;
  final String? restaurantId;
  final List<String> applicableItems;

  Coupon({
    required this.id,
    required this.code,
    required this.description,
    required this.discountType,
    required this.discountValue,
    required this.minPurchase,
    this.maxDiscount,
    required this.startDate,
    required this.endDate,
    required this.isActive,
    required this.usageLimit,
    required this.usedCount,
    this.restaurantId,
    required this.applicableItems,
  });

  factory Coupon.fromJson(Map<String, dynamic> json) {
    return Coupon(
      id: json['_id']?.toString() ?? '',
      code: json['code']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      discountType: json['discountType']?.toString() ?? 'percentage',
      discountValue: (json['discountValue'] as num?)?.toDouble() ?? 0.0,
      minPurchase: (json['minPurchase'] as num?)?.toDouble() ?? 0.0,
      maxDiscount: (json['maxDiscount'] as num?)?.toDouble(),
      startDate: DateTime.parse(json['startDate'].toString()),
      endDate: DateTime.parse(json['endDate'].toString()),
      isActive: json['isActive'] as bool? ?? true,
      usageLimit: json['usageLimit'] as int? ?? 1,
      usedCount: json['usedCount'] as int? ?? 0,
      restaurantId: json['restaurantId']?.toString(),
      applicableItems: (json['applicableItems'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'code': code,
      'description': description,
      'discountType': discountType,
      'discountValue': discountValue,
      'minPurchase': minPurchase,
      'maxDiscount': maxDiscount,
      'startDate': startDate.toIso8601String(),
      'endDate': endDate.toIso8601String(),
      'isActive': isActive,
      'usageLimit': usageLimit,
      'usedCount': usedCount,
      'restaurantId': restaurantId,
      'applicableItems': applicableItems,
    };
  }
}

// Extension for reactive properties
extension ReactiveCoupon on Coupon {
  RxString get rxCode => code.obs;
  RxBool get rxIsActive => isActive.obs;
  RxInt get rxUsedCount => usedCount.obs;
}
